### PR TITLE
fix(ts): add overloads to withAuth middleware

### LINKED ISSUE #7997: withAuth return types do not narrow down to middleware type
## Problem
The `withAuth` middleware returns different values depending on provided arguments. In the current implementation it returns a union such as: `NextMiddlewareWithAuth | Promise<NextMiddlewareResult>`.

When attempting to invoke the return of `withAuth` as the callable middleware, it results in the following error:

```txt
const withAuthMiddleware: Promise<NextMiddlewareResult> | ((request: NextRequestWithAuth, event: NextFetchEvent) => Promise<NextMiddlewareResult>)
This expression is not callable.
  Not all constituents of type 'Promise<NextMiddlewareResult> | ((request: NextRequestWithAuth, event: NextFetchEvent) => Promise<NextMiddlewareResult>)' are callable.
    Type 'Promise<NextMiddlewareResult>' has no call signatures.typescript(2349)
```

## Example
Here is an example of where things breakdown. In this example, we want to extract some outside context from the incoming request, i.e the user's locale, and then attach to the redirect URL as a query parameter.
* NB: this is a very specific application derived from this other [discussion](https://github.com/nextauthjs/next-auth/discussions/6946), but this general pattern could be undertaken for many different reasons

```ts
import { withAuth } from "next-auth/middleware";
import { NextFetchEvent } from "next/server";

// More on how NextAuth.js middleware works: https://next-auth.js.org/configuration/nextjs#middleware

export default async function middleware(req: Request, event: NextFetchEvent) {
  const localeStr = req.headers.get("Accept-Language");
  const locale = localeStr?.split(";")[0].split(",")[0];

  const withAuthMiddleware = withAuth({
    callbacks: {},
    pages: {
      signIn: `/?locale=${locale}`,
    },
  });

  // PROBLEM: Types do not narrow as expected but should be fine. Hover over the function to see its types
  // Fix by @ts-ignore or type cast to NextMiddleware. Leaving as is to showcase a bug.
  return await withAuthMiddleware(req, event);
}

export const config = { matcher: ["/admin", "/me"] };
```

## Solutions?
* Some simple solutions to this include a type cast to `NextMiddleware`, or
* Using `@ts-ignore`,

however I wanted to propose resolving with function overloading on the library side so that things work out of the box! Additionally, the function overloads may may be helpful in surfacing changes to the API to users, compared to the aforementioned ones, which may hide legitimate issues.

PS: I have a potential solution ready which I am testing out and can contribute back to the codebase, but I stopped to open an issue first before proceeding! Thanks.

### How to reproduce ☕️

* [CodeSandbox Link](https://codesandbox.io/p/sandbox/interesting-james-45rtfp?file=%2Fmiddleware.ts%3A13%2C34)

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

### PR DESCRIPTION
**Pull request by rexfordessilfie**
## ☕️ Reasoning

The `withAuth` middleware returns different values depending on provided arguments. In the current implementation it returns a union such as:`NextMiddlewareWithAuth | NextMiddlewareResult | ...`.

This union is type is not narrow enough for use in scenarios where we expect a callable `NextMiddlewareWithAuth` that we can invoke later after getting some outside context from the middleware `request` object. 

In summary, when we want to do something like this:
```ts
export default middleware(req: Request, event: NextFetchEvent){
   // ...do something with Request
   const response = withAuth({...})(req, event)
   // ...do something with Response?
  return response
}

// Side note: This PR also resolves type issues that would arise from doing this instead 
// (for the reason that Request does not extend `NextRequestWithAuth`:
// const response = withAuth(req, {...})
```

**Solution**: This PR adds type overloads such that the types narrow down as expected depending on the arguments supplied to `withAuth`. This lets us do the above without any type issues.

## 🧢 Checklist

- [x] Documentation (N/A)
- [x] Tests (N/A)
- [x] Ready to be merged

## 🎫 Affected issues

Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

Fixes: https://github.com/nextauthjs/next-auth/issues/7997

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by balazsorban44** [comment]
Thanks, could you open this against the `v4` branch? 🙏

**Comment by rexfordessilfie** [comment]
Thanks! Will do.

**Comment by rexfordessilfie** [comment]
Done, @balazsorban44 ✅. Let me know if I did everything right!

**Review by ThangHuuVu — CHANGES_REQUESTED** [review]
Thanks for the detailed bug report and the PR, @rexfordessilfie 🙏 I left a comment, let me know what you think!

File: packages/next-auth/src/next/middleware.ts [hunk 1]
```diff
@@ -186,7 +186,48 @@ export type WithAuthArgs =
...
+  request: Req,
+  event: NextFetchEvent
+) => ReturnType<NextMiddlewareWithAuth>
+
+export function withAuth(
+  middleware: NextMiddlewareWithAuth
+): <Req extends Request = NextRequestWithAuth>(
+  request: Req,
+  event: NextFetchEvent
+) => ReturnType<NextMiddlewareWithAuth>
```
[the remark this answers is not in the document]
Comment (rexfordessilfie) [reply packages/next-auth/src/next/middleware.ts:219]: Thanks for the feedback! Yes, you are right. It was a copy directly from `NextMiddlewareWithAuth`.

I did not consider using `req: NextRequestWithAuth` directly. I noticed I was getting type errors when it was `req: Request` and so I made a generic version of `NextMiddlewareWithAuth` to fix the type errors!

I will switch to this and request a re-review!

**Review by ThangHuuVu — APPROVED** [review]
LGTM!

**Pull request merged**